_internal/utils/mutate.ts

```diff
@@ -35,130 +49,145 @@ export const internalMutate = async <Data>(
   const revalidate = options.revalidate !== false
   const rollbackOnError = options.rollbackOnError !== false
 
-  // Serialize key
-  const [key] = serialize(_key)
-  if (!key) return
+  if (!_key) return
 
-  const [get, set] = createCacheHelper<
-    Data,
-    State<Data, any> & {
-      // The previously committed data.
-      _c?: Data
-    }
-  >(cache, key)
-  const [EVENT_REVALIDATORS, MUTATION, FETCH] = SWRGlobalState.get(
-    cache
-  ) as GlobalState
-
-  const revalidators = EVENT_REVALIDATORS[key]
-  const startRevalidate = () => {
-    if (revalidate) {
-      // Invalidate the key by deleting the concurrent request markers so new
-      // requests will not be deduped.
-      delete FETCH[key]
-      if (revalidators && revalidators[0]) {
-        return revalidators[0](revalidateEvents.MUTATE_EVENT).then(
-          () => get().data
-        )
-      }
+  const KEYS = (SWRGlobalState.get(cache) as GlobalState)[6]
+
+  // If 2nd arg is key filter, return the mutation results of filtered keys
+  if (isFunction(_key)) {
+    const keyFilter = _key
+    const matchedKeys = []
+    for (const k of KEYS.values()) {
+      if (keyFilter(k)) matchedKeys.push(k)
```

Comment:
One nice thing is to catch potential mistakes that people forget to update their code for this breaking change:

```suggestion
      const shouldMutateKey = keyFilter(k)
      if (typeof shouldMutateKey == 'boolean') {
        if (shouldMutateKey) {
          matchedKeys.push(k)
        }
      } else {
        throw new Error('It\'s not allowed to use function to return a key in `mutate` since SWR 2.0.')
      }
```

Reply:
We also have falsy value support for the swr keys, and this feel like we're doing a part of type validation with js. And the return type will become uncertain, if there's no error, it will resolve an array of values or errors, in this case will only bail with 1 error.

Reply:
let's improve the error/warning later as enhancement